src/os/bluestore/BlueFS.cc

```diff
@@ -2729,42 +2756,18 @@ int BlueFS::_flush_range(FileWriter *h, uint64_t offset, uint64_t length)
       ceph_abort_msg("bluefs enospc");
       return r;
     }
-    must_dirty = true;
+    h->file->is_dirty = true;
```

Comment:
Shouldn't we remove h->file->dirty_seq from dirty_files as this point?
Isn't the following scenario error-prone?
- assume h1->file->dirty_seq != log_seq + 1 && h1->file->dirty_seq != 0, i.e. file is already present in dirty_files
- another  _flush_range reaches the same h1->file - is_dirty flag is set, write to slow device is being performed.
- the same scenario with h2 fsync follows and causes replay log saving. which in turn writes newer h1 and causes the same issues as the original one

Reply:
_fsync and _flush_range are executed under the same lock.
_flush_range is no longer dirting files.

If _fsync executes first, then _flush_range will see cleared h->file->is_dirty flag.
If _flush_range executes first, then _fsync ( by _signal_dirty_to_log) will redirty file.
I don't see problem. Did I miss some detail?

Reply:
Although while leaving my previous comment I missed that non-empty dirty_files is rather short living now (as _fsync both propagates and immediately consumes it most of the time) _fsync still might temporary release the lock. This happens while waiting on log_cond in _flush_and_sync_log. Hence hypothetically competing _flush_range might jump in and initiate another write. Then _flush_and_sync_log takes the control back and proceeds with the log write - which might result in the same unsafe log and file data discrepancy.

I realize that the above scenario is rather hypothetical since it implies new _flush_range for the same handle to be called before previous fsync completion. Hopefully RocksDB isn't that crazy.. Just to let you know ;)

And initially I was presuming a bit different scenario. Which doesn't make sense due to exclusive locking and short-living of dirty_files content.